content/_code-samples/trade-in-the-decentralized-exchange/py/trade-in-the-dex.py

```diff
@@ -0,0 +1,243 @@
+
+        # Look up Offers. -----------------------------------------------------------
+        # To buy TST, look up Offers where "TakerGets" is TST:
+        orderbook_info = await client.request(
+            BookOffers(
+                taker=wallet.classic_address,
+                ledger_index="current",
+                taker_gets=we_want["currency"],
+                taker_pays=we_spend["currency"],
+                limit=10,
+            )
+        )
+        print(f"Orderbook:\n{pprint.pformat(orderbook_info.result)}")
+
+        # Estimate whether a proposed Offer would execute immediately, and...
+        # If so, how much of it? (Partial execution is possible)
+        # If not, how much liquidity is above it? (How deep in the order book would
+        # other Offers have to go before ours would get taken?)
+        # Note: These estimates can be thrown off by rounding if the token issuer
+        # uses a TickSize setting other than the default (15). In that case, you
+        # can increase the TakerGets amount of your final Offer to compensate.
+
+        offers = orderbook_info.result.get("offers", [])
+        want_amt = Decimal(we_want["value"])
+        running_total = Decimal(0)
+        if len(offers) == 0:
+            print("No Offers in the matching book. Offer probably won't execute immediately.")
+        else:
+            for o in offers:
+                if Decimal(o["quality"]) <= proposed_quality:
+                    print(f"Matching Offer found, funded with {o.get('owner_funds')} "
+                          f"{we_want['currency']}")
+                    running_total += Decimal(o.get("owner_funds", Decimal(0)))
+                    if running_total >= want_amt:
+                        print("Full Offer will probably fill")
+                        break
+                else:
+                    # Offers are in ascending quality order, so no others after this
+                    # will match either
+                    print("Remaining orders too expensive.")
+                    break
+
+            print(f"Total matched: {min(running_total, want_amt)} {we_want['currency']}")
+            if 0 < running_total < want_amt:
+                print(f"Remaining {want_amt - running_total} {we_want['currency']} "
+                      "would probably be placed on top of the order book.")
+
+        if running_total == 0:
+            # If part of the Offer was expected to cross, then the rest would be placed
+            # at the top of the order book. If none did, then there might be other
+            # Offers going the same direction as ours already on the books with an
+            # equal or better rate. This code counts how much liquidity is likely to be
+            # above ours.
+            #
+            # Unlike above, this time we check for Offers going the same direction as
+            # ours, so TakerGets and TakerPays are reversed from the previous
+            # book_offers request.
+
```

Comment:
Same here, we should always prompt user when an async function is running to let them know what's going on